Firestore/core/src/firebase/firestore/util/filesystem.h

```diff
@@ -70,6 +71,55 @@ Status RecursivelyDelete(const Path& path);
  */
 Path TempDir();
 
+/**
+ * Implements an iterator over the contents of a directory. Initializes to the
+ * first entry in the directory.
+ */
+class DirectoryIterator {
+ public:
+  /**
+   * `path` should outlive the iterator.
+   */
+  explicit DirectoryIterator(const Path& path);
+  ~DirectoryIterator();
+
+  DirectoryIterator(const DirectoryIterator& other) = delete;
+
+  DirectoryIterator& operator=(const DirectoryIterator& other) = delete;
+
+  /**
+   * Advances the iterator.
+   */
+  void Next();
+
+  /**
+   * Returns true if `Next()` and `file()` can be called on the iterator.
+   * If `Valid() == false && status().ok()`, then iteration has finished.
+   */
+  bool Valid();
+
+  /**
+   * Return the full path of the current entry pointed to by the iterator.
+   */
+  Path file();
+
+  /**
+   * Returns the last error encountered by the iterator, or OK.
+   */
+  Status status() {
+    return status_;
+  }
+
+ private:
+  // Internal method to move to the next directory entry
+  void Advance();
+
+  Status status_;
+  DIR* dir_;
```

Comment:
`DIR` and `struct dirent` are both POSIX-specific and won't work well on Windows.

The way to handle this is to declare the contents like so:

```c++
class DirectoryIterator {
  // ...

 private:
  struct Rep;
  std::unique_ptr<Rep> rep_;
};
```

Now in `filesystem_posix.cc`, add the following:

```c++
struct DirectoryIterator::Rep {
  DIR* dir;
  struct dirent* entry;
};
```

This works because the standard specifically provides for `std::unique_ptr` to take as parameters types that are only forward declared.

Reply:
Done.